rfcs/text/0010_service_status.md

```diff
@@ -0,0 +1,337 @@
+     */
+    unavailableWhen<P, Q, B>(
+      predicate: ServiceStatusLevel |
+        (self: ServiceStatus, core: CoreStatus, plugins: Record<string, ServiceStatus>) => boolean,
+      handler: RouteHandler<P, Q, B>
+    ): RouteHandler<P, Q, B>;
+  }
+}
+```
+
+## Additional Examples
+
+### Combine inherited status with check against external dependency
+```ts
+const getExternalDepHealth = async () => {
+  const resp = await window.fetch('https://myexternaldep.com/_healthz');
+  return resp.json();
+}
+
+// Create an observable that checks the status of an external service every every 10s 
+const myExternalDependency$: Observable<ServiceStatusLevel> = interval(10000).pipe(
+  mergeMap(() => of(getExternalDepHealth())),
+  map(health => health.ok ? ServiceStatusLevel.available : ServiceStatusLevel.unavailable),
+  catchError(() => of(ServiceStatusLevel.unavailable))
+);
+
+// Merge the inherited status with the external check
+core.status.set(
+  combineLatest(
+    core.status.inherited$,
+    myExternalDependency$
+  ).pipe(
+    map(([inherited, external]) => ({
+      level: Math.max(inherited.level, external)
+    }))
+  )
+);
+```
+
+# Drawbacks
+
+1. **The default behaviors and inheritance of statuses may appear to be "magic" to developers who do not read the documentation about how this works.** Compared to the legacy status mechanism, these defaults are much more opinionated and the resulting status is less explicit in plugin code compared to the legacy `mirrorPluginStatus` mechanism.
+2. **The default behaviors and inheritance may not fit real-world status very well.** If many plugins must customize their status in order to opt-out of the defaults, this would be a step backwards from the legacy mechanism.
+
+# Alternatives
+
+We could somewhat reduce the complexity of the status inheritance by leveraging the dependencies between plugins to enable and disable plugins based on whether or not their upstream dependencies are available. This may simplify plugin code but would greatly complicate how Kibana fundamentally operates, requiring that plugins may get stopped and started multiple times within a single Kibana server process. We would be trading simplicity in one area for complexity in another.
+
+# Adoption strategy
+
+By default, most plugins would not need to do much at all. Today, very few plugins leverage the legacy status system. The majority of ones that do, simply call the `mirrorPluginStatus` utility to follow the status of the legacy elasticsearch plugin.
+
+Plugins that wish to expose more detail about their availability will easily be able to do so, including providing detailed information such as links to documentation to resolve the problem.
+
+Depending on [Drawback #3](#drawbacks), we may need to break the existing status API, however this seems unlikely. This could have a large impact on users who use this API for checking Kibana node health in load balancer configurations.
```

Comment:
> Dependeng on Drawback #3

Looks like drawback 3 was removed

Reply:
Ah good catch. Need to update this paragraph.